packages/world/src/modules/core/implementations/WorldRegistrationSystem.sol

```diff
@@ -62,27 +62,46 @@ contract WorldRegistrationSystem is System, IWorldErrors {
    * If the namespace doesn't exist yet, it is registered.
    * The system is granted access to its namespace, so it can write to any table in the same namespace.
    * If publicAccess is true, no access control check is performed for calling the system.
+   *
+   * Note: this function doesn't check whether a system already exists at the given selector,
+   * making it possible to upgrade systems.
    */
   function registerSystem(bytes32 resourceSelector, WorldContextConsumer system, bool publicAccess) public virtual {
     // Require the name to not be the namespace's root name
     if (resourceSelector.getName() == ROOT_NAME) revert InvalidSelector(resourceSelector.toString());
 
-    // Require the system to not exist yet
-    if (SystemRegistry.get(address(system)) != 0) revert SystemExists(address(system));
+    // Require this system to not be registered at a different resource selector yet
+    bytes32 existingResourceSelector = SystemRegistry.get(address(system));
+    if (existingResourceSelector != 0 && existingResourceSelector != resourceSelector) {
+      revert SystemExists(address(system));
+    }
 
     // If the namespace doesn't exist yet, register it
     // otherwise require caller to own the namespace
     bytes16 namespace = resourceSelector.getNamespace();
     if (ResourceType.get(namespace) == Resource.NONE) registerNamespace(namespace);
     else AccessControl.requireOwnerOrSelf(namespace, _msgSender());
 
-    // Require no resource to exist at this selector yet
-    if (ResourceType.get(resourceSelector) != Resource.NONE) {
+    // Require no resource other than a system to exist at this selector yet
+    Resource resourceType = ResourceType.get(resourceSelector);
+    if (resourceType != Resource.NONE && resourceType != Resource.SYSTEM) {
       revert ResourceExists(resourceSelector.toString());
     }
 
-    // Store the system resource type
-    ResourceType.set(resourceSelector, Resource.SYSTEM);
+    // Check if a system already exists at this resource selector
+    address existingSystem = Systems.getSystem(resourceSelector);
+
+    // If there is an existing system with this resource selector, remove it
+    if (existingSystem != address(0)) {
+      // Remove the existing system from the system registry
+      SystemRegistry.deleteRecord(existingSystem);
+
+      // Remove the existing system's access to its namespace
+      ResourceAccess.deleteRecord(namespace, existingSystem);
+    } else {
+      // Otherwise, this is a new system, so register its resource type
+      ResourceType.set(resourceSelector, Resource.SYSTEM);
+    }
```

Comment:
idle thought: I wonder if it's worth encoding the resource type into the selector, so selectors become something like
```
type (bytes2) + namespace (bytes14) + name (bytes16)
```
this would save a little on storage (but maybe not in any meaningful way if it's only used during registration) and could help during introspection (it's part of the selector/identifier itself rather than a separate lookup)

Reply:
agree that might make sense!